haystack_experimental/components/extractors/llm_metadata_extractor.py

```diff
@@ -0,0 +1,257 @@
+
+    def _check_prompt(self):
+        if self.input_text not in self.prompt:
+            raise ValueError(f"{self.input_text} must be in the prompt.")
+
+    @staticmethod
+    def _init_generator(generator_api: LLMProvider, generator_api_params: Dict[str, Any]):
+        """
+        Initialize the chat generator based on the specified API provider and parameters.
+        """
+        if generator_api == LLMProvider.OPENAI:
+            return OpenAIGenerator(**generator_api_params)
+        if generator_api == LLMProvider.OPENAI_AZURE:
+            return AzureOpenAIGenerator(**generator_api_params)
+        if generator_api == LLMProvider.AWS_BEDROCK:
+            amazon_bedrock_generator.check()
+            return AmazonBedrockGenerator(**generator_api_params)
+        if generator_api == LLMProvider.GOOGLE_VERTEX:
+            vertex_ai_gemini_generator.check()
+            return VertexAIGeminiGenerator(**generator_api_params)
+        raise ValueError(f"Unsupported generator API: {generator_api}")
+
+    def is_valid_json_and_has_expected_keys(self, expected: List[str], received: str) -> bool:
+        """
+        Output must be a valid JSON with the expected keys.
+
+        :param expected:
+            Names of expected outputs
+        :param received:
+            Names of received outputs
+
+        :raises ValueError:
+            If the output is not a valid JSON with the expected keys:
+            - with `raise_on_failure` set to True a ValueError is raised.
+            - with `raise_on_failure` set to False a warning is issued and False is returned.
+
+        :returns:
+            True if the received output is a valid JSON with the expected keys, False otherwise.
+        """
+        try:
+            parsed_output = json.loads(received)
+        except json.JSONDecodeError:
+            msg = "Response from LLM evaluator is not a valid JSON."
+            if self.raise_on_failure:
+                raise ValueError(msg)
+            warn(msg)
+            logger.warning(msg)
+            return False
+
+        if not all(output in parsed_output for output in expected):
+            msg = f"Expected response from LLM evaluator to be JSON with keys {expected}, got {received}."
+            if self.raise_on_failure:
+                raise ValueError(msg)
+            warn(msg)
+            logger.warning(msg)
+            return False
+
+        return True
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Serializes the component to a dictionary.
+
+        :returns:
+            Dictionary with serialized data.
+        """
+        return default_to_dict(
+            self,
+            prompt=self.prompt,
+            input_text=self.input_text,
+            expected_keys=self.expected_keys,
+            raise_on_failure=self.raise_on_failure,
+            generator_api=self.generator_api.value,
+            generator_api_params=self.generator_api_params,
```

Comment:
Out of curiosity don't we need to handle the api key specifically when converting `to_dict`? In our generators we typically would call 
```python
api_key=self.api_key.to_dict()
```

Perhaps instead of doing this we could get all api_params by doing something like
```
llm_dict = llm_provider.to_dict()
generator_api_params = llm_dict["init_params"]
```
this way we could leverage the `to_dict` and `from_dict` methods of the underlying generator which handles the sede of the API key secret right?

Reply:
You are right, we are not handling the key, assuming the user will use the default one.

The question is whether we want to serialise an LLM provider that has already been initialised or simply the `generator_api` of the LLM provider and the `generator_api_params`, and then the `from_dict()` will use this information to initialise it when loaded.

Your approach would tie the `LLMMetadataExtractor` to an LLM provider when serialised/deserialised.

You are the main user @sjrl :) Let me know how you prefer it.